### ISSUE TITLE
inconsistent result of ```tf.image.adjust_hue``` on CPU and GPU

### ISSUE THREAD

**Issue by usmonali4** [tensorflow/tensorflow#86256]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

2.18

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

3.11

### CUDA/cuDNN version

12.5/9

### GPU model and memory

Tesla T4

### Current behavior?

getting inconsistent results of ```tf.image.adjust_hue``` between CPU and GPU

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

images = tf.constant([
    [[ 1.9720840,  2.1302242, -0.1902120],
     [ 0.6557856, -1.3016001,  1.1452782]],
    
    [[-2.2193234,  0.3198028,  0.9568117],
     [-0.3937407, -0.0503466, -0.3693791]]
], dtype=tf.float32)

delta = tf.constant(-0.7441734, dtype=tf.float32)

with tf.device('CPU:0'):
    adjusted_cpu = tf.image.adjust_hue(images, delta)
    print("Adjusted Hue on CPU:\n", adjusted_cpu)

with tf.device('GPU:0'):
    adjusted_gpu = tf.image.adjust_hue(images, delta)
    print("Adjusted Hue on GPU:\n", adjusted_gpu)


is_consistent = tf.experimental.numpy.allclose(adjusted_cpu, adjusted_gpu, atol=1e-5, rtol=1e-6)

max_abs_diff = tf.reduce_max(tf.abs(adjusted_cpu - adjusted_gpu)).numpy()

print("Max absolute difference:", max_abs_diff)
print("Consistency check (CPU vs GPU) with atol=1e-5 and rtol=1e-6:", is_consistent.numpy())
```

### Relevant log output

```shell
Adjusted Hue on CPU:
 tf.Tensor(
[[[-0.190212    2.1302242   1.2092681 ]
  [ 1.1452782  -0.48211157 -1.3016001 ]]

 [[ 0.11679006 -2.2193234   0.9568117 ]
  [-0.3937407  -0.25841027 -0.0503466 ]]], shape=(2, 2, 3), dtype=float32)

Adjusted Hue on GPU:
 tf.Tensor(
[[[-0.19021209  2.1302242   1.209268  ]
  [ 1.1452781  -0.48211193 -1.3016001 ]]

 [[ 0.11678863 -2.2193234   0.95681167]
  [-0.0503466  -0.0503466  -0.0503466 ]]], shape=(2, 2, 3), dtype=float32)

Max absolute difference: 0.3433941
Consistency check (CPU vs GPU) with atol=1e-5 and rtol=1e-6: False
```

**Comment by rileythompso** [comment]
The discrepancy you're observing between CPU and GPU outputs when using `tf.image.adjust_hue` is primarily due to differences in floating-point precision and the underlying implementations on these platforms. Such variations are common in numerical computations across different hardware architectures.

**Potential Causes:**

1. **Floating-Point Precision:** CPUs and GPUs may handle floating-point arithmetic differently, leading to minor discrepancies in results.

2. **Implementation Differences:** The algorithms or optimizations used for image processing operations can vary between CPU and GPU implementations, contributing to inconsistent outputs.

**Recommended Solutions:**

1. **Set a Global Seed:** Ensure reproducibility by setting a global random seed.

   ```python
   import tensorflow as tf
   tf.random.set_seed(42)
   ```

2. **Use CPU Consistently:** If consistency is crucial, consider performing the hue adjustment exclusively on the CPU.

   ```python
   with tf.device('/CPU:0'):
       adjusted_image = tf.image.adjust_hue(image, delta)
   ```

3. **Adjust Tolerance Levels:** When comparing CPU and GPU results, allow for minor differences by setting appropriate tolerance levels.

   ```python
   is_consistent = tf.experimental.numpy.allclose(adjusted_cpu, adjusted_gpu, atol=1e-5, rtol=1e-6)
   ```

4. **Preprocess Data:** Normalize or preprocess your images to minimize the impact of precision differences.

   ```python
   image = tf.image.convert_image_dtype(image, tf.float32)
   ```

**Additional Considerations:**

- **Library Versions:** Ensure that your TensorFlow and CUDA versions are compatible and up-to-date.

- **Hardware Variations:** Different GPU models may exhibit varying behaviors due to architectural differences.

By implementing these strategies, you can mitigate the inconsistencies between CPU and GPU computations in TensorFlow.

**Comment by tilakrayal** [comment]
@usmonali4,
I think the issue is specific to GPU, I was able to reproduce the issue on colab with GPU runtime. But, when the inputs are changed to float64 precision, the results are as expected. 
The reason could be due to the optimized way of calculating numbers on Nvidia which is different compared to others, there will be a small amount of precision errors.
The same behavior is not observed on Apple M1, using numpy or in CPU.
https://github.com/tensorflow/tensorflow/issues/58479


Thank you!

**Comment by usmonali4** [comment]
@tilakrayal, 
I also suspect the issue to be GPU specific, and I understand that as api calculations vary across devices, minor differences between results are expected. However, each data type should have an acceptable tolerance range. I.e deviations beyond this range may indicate inadequate accuracy control on certain devices. e.g, current report shows a difference greater than 0.34 for float32 values, which is excessive given float32 precision. Even when considering relaxed tolerance thresholds as atol=1e-1 and rtol=1e-1 the difference would fall outside the range. Glad to hear your thoughts on this; thanks for your time.

**Comment by jobs-git** [comment]
@usmonali4 @tilakrayal 

I got the same issue on a different configuration:

```bash
TensorFlow Version: 2.16.2
Python Version: 3.10.16 | packaged by conda-forge | (main, Dec  5 2024, 14:16:10) [GCC 13.3.0]
Python Implementation: CPython
CUDA Version (TensorFlow): 12.0
cuDNN Version (TensorFlow): 8
GPU Name (PyTorch): NVIDIA GeForce RTX 4070 SUPER
CUDA Version (PyTorch): 12.0
cuDNN Version (PyTorch): 8907
Driver Version: 550.144.03
Ubuntu Version: PRETTY_NAME="Ubuntu 24.04.1 LTS"
Model name:                           AMD Ryzen 5 7600 6-Core Processor
Adjusted Hue on CPU:
 tf.Tensor(
[[[-0.190212    2.1302242   1.2092681 ]
  [ 1.1452782  -0.48211157 -1.3016001 ]]

 [[ 0.11679006 -2.2193234   0.9568117 ]
  [-0.3937407  -0.25841027 -0.0503466 ]]], shape=(2, 2, 3), dtype=float32)
Adjusted Hue on GPU:
 tf.Tensor(
[[[-0.19021201  2.1302242   1.2092681 ]
  [ 1.1452782  -0.4821118  -1.3016    ]]

 [[ 0.11678863 -2.2193234   0.95681167]
  [-0.0503466  -0.0503466  -0.0503466 ]]], shape=(2, 2, 3), dtype=float32)
Max absolute difference: 0.3433941
Consistency check (CPU vs GPU) with atol=1e-5 and rtol=1e-6: False
```

**Comment by avmohansai** [comment]
Hi @usmonali4  ,
Thank you so much for filing this issue. We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

From the example shared, this does not appear to be a typical minor floating point precision difference between CPU and GPU executions. While small numerical deviations are generally expected across hardware backends, the reproducer shows a significantly larger mismatch in one output pixel, where the GPU result collapses all three channels to the same value `([-0.0503466, -0.0503466, -0.0503466])` while the CPU output produces distinct channel values.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
